converter/converters/pod/affinity.go

```diff
@@ -0,0 +1,398 @@
+		case "=":
+			op = metav1.LabelSelectorOpIn
+		case "!=":
+			op = metav1.LabelSelectorOpNotIn
+		default:
+			glog.Fatal("unreachable")
+		}
+		reqs = append(reqs, metav1.LabelSelectorRequirement{
+			Key:      expr.Key,
+			Operator: op,
+			Values:   expr.Values,
+		})
+	}
+
+	if len(labels) == 0 {
+		labels = nil
+	}
+
+	if len(reqs) == 0 {
+		reqs = nil
+	}
+
+	return &v1.PodAffinityTerm{
+		LabelSelector: &metav1.LabelSelector{
+			MatchLabels:      labels,
+			MatchExpressions: reqs,
+		},
+	}, nil
+}
+
+func nodeAffinity(as []string) (*v1.NodeAffinity, error) {
+	if len(as) == 0 {
+		return nil, nil
+	}
+
+	hard, soft, err := nodeAffinities(as)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hard) == 0 {
+		hard = nil
+	}
+
+	if len(soft) == 0 {
+		soft = nil
+	}
+
+	return &v1.NodeAffinity{
+		RequiredDuringSchedulingIgnoredDuringExecution:  &v1.NodeSelector{hard},
+		PreferredDuringSchedulingIgnoredDuringExecution: soft,
+	}, nil
+}
+
+func nodeAffinities(as []string) (hard []v1.NodeSelectorTerm, soft []v1.PreferredSchedulingTerm, err error) {
+	hard = []v1.NodeSelectorTerm{}
+	soft = []v1.PreferredSchedulingTerm{}
+	for _, a := range as {
+		segs := strings.Split(a, ":")
+		l := len(segs)
+
+		var term *v1.NodeSelectorTerm
+		if l < 1 {
+			err = fmt.Errorf("unrecognized NodeAffinity term %s", a)
+			return
+		} else {
+			term, err = nodeExprs(segs[0])
+			if err != nil {
+				return
+			}
+		}
+
+		if l < 2 {
+			hard = append(hard, *term)
+			continue
+		} else {
+			if segs[1] != "soft" {
+				err = fmt.Errorf("unrecognized NodeAffinity term %s", a)
+				return
+			}
+		}
+
+		var weight int64
+		if l < 3 {
+			weight = 1
+		} else {
+			weight, err = strconv.ParseInt(segs[2], 10, 32)
+			if err != nil {
+				return
+			}
+		}
+
+		soft = append(soft, v1.PreferredSchedulingTerm{
+			Weight:     int32(weight),
+			Preference: *term,
+		})
+	}
+
+	return
+}
+
+func nodeExprs(s string) (*v1.NodeSelectorTerm, error) {
+	reqs := []v1.NodeSelectorRequirement{}
+	segs := strings.Split(s, "&")
+	for _, seg := range segs {
+		expr, err := parseExpr(seg, []string{"!=", "=", ">", "<"})
+		if err != nil {
+			return nil, err
+		}
+
+		if expr == nil {
+			if seg[0] == '!' {
+				reqs = append(reqs, v1.NodeSelectorRequirement{
+					Key:      seg[1:],
+					Operator: v1.NodeSelectorOpDoesNotExist,
+				})
+			} else {
+				reqs = append(reqs, v1.NodeSelectorRequirement{
+					Key:      seg,
+					Operator: v1.NodeSelectorOpExists,
+				})
+			}
+
+			continue
+		}
+
+		var op v1.NodeSelectorOperator
+		switch expr.Op {
+		case "=":
+			op = v1.NodeSelectorOpIn
+		case "!=":
+			op = v1.NodeSelectorOpNotIn
+		case ">":
+			op = v1.NodeSelectorOpGt
+		case "<":
+			op = v1.NodeSelectorOpLt
+		default:
+			glog.Fatal("unreachable")
+		}
+		reqs = append(reqs, v1.NodeSelectorRequirement{
+			Key:      expr.Key,
+			Operator: op,
+			Values:   expr.Values,
+		})
+	}
+
+	return &v1.NodeSelectorTerm{reqs}, nil
+}
+
+type Expr struct {
```

Comment:
use v1 types just parse it without converting types unnecessarily

Reply:
added a comment about using Expr as an AST for two separate v1 types.